rye/src/cli/publish.rs

```diff
@@ -257,24 +451,171 @@ fn get_trimmed_user_input() -> Result<String, Error> {
+        )
+        .unwrap();
+
+        assert_eq!(credentials.username, Some(cli_username));
+        assert_eq!(
+            credentials.password.unwrap().expose_secret(),
+            cli_token.expose_secret()
+        );
+        assert_eq!(repository.url, Some(cli_repo_url));
+    }
+
+    #[test]
+    fn test_repository_config_resolution_file_and_cli() {
+        let mut credentials_table = Item::Table(Table::new());
+        credentials_table["repository-url"] =
+            Item::Value("https://test.pypi.org/".to_string().into());
+        credentials_table["username"] = Item::Value("username".to_string().into());
+
+        let cli_repo = "pypi".to_string();
+        let cli_repo_url = None;
+        let cli_username = None;
+        let cli_token = Secret::new("token".to_string());
+
+        let config = resolve_credentials(Some(&credentials_table), cli_username, Some(&cli_token));
+        let repository =
+            resolve_repository(Some(&credentials_table), Some(cli_repo), cli_repo_url).unwrap();
+
+        assert_eq!(
+            config.password.unwrap().expose_secret(),
+            cli_token.expose_secret()
+        );
+        assert_eq!(
+            repository.url.unwrap().to_string(),
+            "https://test.pypi.org/".to_string()
+        );
+    }
+
+    #[test]
+    fn test_repository_config_resolution_keyring_fallback() {
+        let credentials_table = Item::Table(Table::new());
+
+        let cli_repo = "pypi".to_string();
+        let cli_repo_url = Url::parse("https://test.pypi.org/").unwrap();
+        let cli_username = None;
+        let cli_token = None;
+
+        let credentials = resolve_credentials(Some(&credentials_table), cli_username, cli_token);
+        let repository = resolve_repository(
+            Some(&credentials_table),
+            Some(cli_repo),
+            Some(cli_repo_url.clone()),
+        )
+        .unwrap();
+
+        assert!(credentials.username.is_none());
+        assert!(credentials.password.is_none());
+        assert_eq!(repository.url.unwrap(), cli_repo_url);
+    }
+
+    #[test]
+    fn test_save_rye_credentials() {
+        let _tempdir = tempdir().unwrap().path();
+        let mut _credentials_table = Document::new();
+
+        let _cli_repo = "pypi".to_string();
+        let _cli_repo_url = Url::parse("https://test.pypi.org/").unwrap();
+        let _cli_username = "username".to_string();
+        let _cli_token = Secret::new("token".to_string());
+
+        todo!(); // RYE_HOME
+
+        // save_rye_credentials(
+        //     &mut credentials_table,
+        //     &Credentials {
+        //         username: Some("username".to_string()),
+        //         password: Some(cli_token),
+        //     },
+        //     &Repository {
+        //         name: Some(cli_repo),
+        //         url: Some(cli_repo_url),
+        //     },
+        //     false,
+        //     None,
+        // )
+        // .unwrap();
+    }
```

Comment:
This will fail. I'm hoping to mark this as ready either later tonight or tomorrow at some point. Would be cool to have prompt testing with `rexpect` or something.